### PR TITLE
Make send report message from discord a celery task.

### LINKED ISSUE #2992: Migrate flagged message Discord notifications to Celery task
Follow-up to #2641

### PR DESCRIPTION
**Pull request by espoirMur**
Fixes #2992

### CODE REVIEW

**Comment by espoirMur** [comment]
I have done most of what was required for the ticket.. the next step will be how to test if this is working.

If anyone has idea  on how to test this, please do let me know..   @olliestanley

**Comment by olliestanley** [comment]
> I have done most of what was required for the ticket.. the next step will be how to test if this is working.
> 
> If anyone has idea on how to test this, please do let me know.. @olliestanley

If you run `docker compose --profile ci up --build --attach-dependencies` it will run the backend and website locally including the changes you have made. You can use this for testing. Alternatively you could run the services separately

File: .python-version [hunk 1]
```diff
@@ -1 +1 @@
-3.10.8
+3.10.7
```
Comment (olliestanley) [inline .python-version:1]: Surely we don't need to change our Python version for this?

File: README.md [hunk 2]
```diff
@@ -121,8 +121,8 @@ addition to `--profile ci` in the above command.
 
 ## The Vision
 
-We are not going to stop at replicating ChatGPT. We want to build the assistant
-of the future, able to not only write email and cover letters, but do meaningful
+We are not going to stop replicating ChatGPT. We want to build the assistant of
```
Comment (olliestanley) [inline README.md:124]: Can we revert these changes? "stop at replicating" is correct here, not "stop replicating" and the original text is fine

File: backend/oasst_backend/prompt_repository.py [hunk 3]
```diff
@@ -39,6 +39,9 @@
 from sqlalchemy.orm.attributes import flag_modified
 from sqlmodel import JSON, Session, and_, func, literal_column, not_, or_, text, update
 
+# TODO: Add doc string to all methods and classes.
```
Comment (olliestanley) [inline backend/oasst_backend/prompt_repository.py:42]: I don't think we should have random TODOs like this, if we want to embark on adding docstrings throughout the code we should raise a GitHub issue for it instead

File: backend/oasst_backend/scheduled_tasks.py [hunk 4]
```diff
@@ -15,6 +15,8 @@
 from oasst_shared.utils import utcnow
 from sqlmodel import select
 
+# TODO: move each task to a separate file in the future to avoid cluttering this file
```
Comment (olliestanley) [inline backend/oasst_backend/scheduled_tasks.py:18]: I also don't think this should be a TODO, for now we have very few tasks so one file is file and probably will be for a long time

File: backend/requirements.txt [hunk 5]
```diff
@@ -14,8 +14,8 @@ pydantic==1.10.4
 pydantic[email]==1.10.4
 python-dotenv==0.21.0
 python-jose[cryptography]==3.3.0
-redis
-requests
+redis==4.5.5
+requests==2.30.0
```
Comment (olliestanley) [inline backend/requirements.txt:18]: Are these requirement versions needed for the new functionality to work?

File: backend/requirements_worker.txt [hunk 6]
```diff
@@ -13,7 +13,8 @@ pydantic==1.9.1
 pydantic[email]==1.9.1
 python-dotenv==0.21.0
 python-jose[cryptography]==3.3.0
-redis
+redis==4.5.5
+requests==2.30.0
```
Comment (olliestanley) [inline backend/requirements_worker.txt:17]: As above, are these requirement versions needed for the new functionality to work?

File: docker-compose.yaml [hunk 7]
```diff
@@ -117,7 +117,7 @@ services:
     build:
       dockerfile: docker/Dockerfile.backend-worker
       context: .
-    command: celery -A oasst_backend.celery_worker worker -l info
+    command: celery -A oasst_backend.celery_worker worker -l info -E
```
Comment (olliestanley) [inline docker-compose.yaml:120]: What is the `-E` flag for here?

File: backend/oasst_backend/prompt_repository.py [hunk 8]
```diff
@@ -547,7 +549,7 @@ def store_text_labels(self, text_labels: protocol_schema.TextLabels) -> tuple[Te
                         message_id, protocol_schema.EmojiOp.add, protocol_schema.EmojiCode.red_flag
                     )
 
-                    discord.send_new_report_message(message=message, label_text=text_labels.text, user_id=self.user_id)
+                    send_new_report_message.delay(message=message, label_text=text_labels.text, user_id=self.user_id)
```
Comment (olliestanley) [inline backend/oasst_backend/prompt_repository.py:552]: So I think what we want to do here is add this to a queue (maybe in the backend Redis) and then run the Celery task on a timer, which will send all recently flagged messages as a single Discord message, this way we can avoid spamming Discord with API calls on every flag

File: backend/.env.example [hunk 9]
```diff
@@ -1 +1,9 @@
 HUGGING_FACE_API_KEY=HF API KEY
+DATABASE_URI="postgresql://<username>:<password>@<host>/<database_name>"
+BACKEND_CORS_ORIGINS=["http://localhost", "http://localhost:4200", "http://localhost:3000", "http://localhost:8080", "https://localhost", "https://localhost:4200", "https://localhost:3000", "https://localhost:8080", "http://dev.oasst.laion.ai", "https://stag.oasst.laion.ai", "https://oasst.laion.ai"]
+REDIS_HOST=localhost
+REDIS_PORT=6379
+
+
+export DEBUG_SKIP_EMBEDDING_COMPUTATION=False
+export DEBUG_SKIP_TOXICITY_CALCULATION=False
```
Comment (olliestanley) [inline backend/.env.example:9]: Thanks for adding all these! This is much cleaner than having the examples in the readme

File: backend/oasst_backend/prompt_repository.py (hunk 8 above)
Comment (espoirMur) [reply backend/oasst_backend/prompt_repository.py:552]: I see this one, so you mean making this a scheduled task? 

Do you mean having a schedule when to run them? if that is the case do you have an ideas about when we should run them?

File: docker-compose.yaml (hunk 7 above)
Comment (espoirMur) [reply docker-compose.yaml:120]: The -E flag hel to list all the task in the worker

File: backend/requirements_worker.txt (hunk 6 above)
Comment (espoirMur) [reply backend/requirements_worker.txt:17]: Those are the latest version of the requirements, I added them to have the requirements version pinned in the file.

File: backend/requirements.txt (hunk 5 above)
Comment (espoirMur) [reply backend/requirements.txt:18]: check the below comment..

File: backend/oasst_backend/scheduled_tasks.py (hunk 4 above)
Comment (espoirMur) [reply backend/oasst_backend/scheduled_tasks.py:18]: Okay, I can remove that..

File: README.md (hunk 2 above)
Comment (espoirMur) [reply README.md:124]: Yes, I can do that.. That came from my auto-correct..

File: backend/oasst_backend/prompt_repository.py (hunk 8 above)
Comment (olliestanley) [reply backend/oasst_backend/prompt_repository.py:552]: Yes. I think a fairly short schedule (e.g. once every 10 minutes) would be fine, we just want to avoid constant spam of Discord messages

Comment (espoirMur) [reply backend/oasst_backend/prompt_repository.py:550]: I kinda see, what you mean.

We just need to push those messages into the queue, and then at a given time, fetch the flagged message and send them in Discord.

**Comment by andreaskoepf** [comment]
@espoirMur could you resolve the conflicts please?

**Review by andreaskoepf — APPROVED** [review]
Thanks a lot, also for cleanup of readme! :-)
